crates/turborepo-lib/src/process/child.rs

```diff
@@ -275,57 +284,17 @@ impl Child {
+            let manager = ChildStateManager {
+                shutdown_style,
+                task_state,
+                exit_tx,
+            };
             tokio::select! {
                 command = command_rx.recv() => {
-                    let state = match command {
-                        // we received a command to stop the child process, or the channel was closed.
-                        // in theory this happens when the last child is dropped, however in practice
-                        // we will always get a `Permit` from the recv call before the channel can be
-                        // dropped, and the channel is not closed while there are still permits
-                        Some(ChildCommand::Stop) | None => {
-                            debug!("stopping child process");
-                            shutdown_style.process(&mut child).await
-                        }
-                        // we received a command to kill the child process
-                        Some(ChildCommand::Kill) => {
-                            debug!("killing child process");
-                            ShutdownStyle::Kill.process(&mut child).await
-                        }
-                    };
-
-                    match state {
-                        ChildState::Exited(exit) => {
-                            // ignore the send error, failure means the channel is dropped
-                            exit_tx.send(Some(exit)).ok();
-                        }
-                        ChildState::Running(_) => {
-                            debug_assert!(false, "child state should not be running after shutdown");
-                        }
-                    }
-
-                    {
-                        let mut task_state = task_state.write().await;
-                        *task_state = state;
-                    }
+                    manager.handle_child_command(command, &mut child).await;
                 }
                 status = child.wait() => {
-                    debug!("child process exited normally");
-                    // the child process exited
-                    let child_exit = match status.map(|s| s.code()) {
-                        Ok(Some(c)) => ChildExit::Finished(Some(c)),
-                        // if we hit this case, it means that the child process was killed
-                        // by someone else, and we should report that it was killed
-                        Ok(None) => ChildExit::KilledExternal,
-                        Err(_e) => ChildExit::Failed,
-                    };
-                    {
-                        let mut task_state = task_state.write().await;
-                        *task_state = ChildState::Exited(child_exit);
-                    }
-
-                    // ignore the send error, the channel is dropped anyways
-                    exit_tx.send(Some(child_exit)).ok();
-
+                    manager.handle_child_exit(status).await;
```

Comment:
[learn rust q]: interesting/unintuitive rust/tokio syntax
```rust
foo = something() => {
  // foo also available here
}
```

Reply:
I am in full agreement, I find [`tokio::select`](https://docs.rs/tokio/latest/tokio/macro.select.html) to be unintuitive macro that causes a lot of problems for me.